NDTensors/ext/NDTensorsGPUArraysCoreExt/contract.jl

```diff
@@ -0,0 +1,47 @@
+using Adapt: adapt
+using GPUArraysCore: AbstractGPUArray
+using NDTensors: NDTensors, DenseTensor, DiagTensor, contract!, dense, inds, Tensor
+using NDTensors.Expose: Exposed, expose, unexpose
+using NDTensors.TypeParameterAccessors: parenttype, set_ndims
+
+## In this function we convert the DiagTensor to a dense tensor and
+## Feed it back into contract
+function NDTensors.contract!(
+  output_tensor::Exposed{<:AbstractGPUArray,<:DenseTensor},
+  labelsoutput_tensor,
+  tensor1::Exposed{<:Any,<:DiagTensor},
```

Comment:
One concern I have about making the type signature this general is that this appears to allow contracting Dense GPU tensors with Diag CPU tensors, which I think we don't want to allow right now since we don't allow contracting Dense GPU tensors with Dense CPU tensors (since it isn't a very well defined operation, i.e. should the result end up on GPU or CPU, and also may indicate someone making a mistake).

What happens with the current PR if you contract a Dense GPU tensor with a Diag CPU tensor? Does that end up in this function, go to another function, throw an error, etc.? Probably to be safe we should restrict this signature more and make it only work for uniform Diag tensors, either with:
```julia
tensor1::Exposed{<:Number,<:DiagTensor},
```
or:
```julia
tensor1::Exposed{<:Number,<:UniformDiagTensor},
```
but I'm curious what the behavior is now without making that change.

Reply:
I updated this to the `Exposed{<:Number, <:DiagTensor}` version. Thanks!